### PR TITLE
Add Load Bearing Vests

### PR DESCRIPTION
**Pull request by bombasticSlacks**
#### Summary
Content "Added modular load bearing vests"

#### Purpose of change
Vests to wear above armor are currently very static and boring however there are a number of interesting accessories already in game that are supposed to attach to MOLLE vests and gear. This is an initial implementation of this attaching system. It only currently will work with the added items and will need a bit more work to be able to be abstracted to all MOLLE items. 

#### Describe the solution
Still to do:

- [x] fix item groups now that I changed ids 
- [x] add items to army surplus and maybe military bunkers. 
- [x] fix an issue with stacking items and removing them from the vests. ( I had migration pockets in the wrong place)

This PR does 6 separate things:

1. Adds 3 new wearable vests: [Light Load Bearing Vest](https://www.amazon.com/OAREA-Tactical-Assault-Multicam-Multi-Pockets/dp/B07LF8F1P1/ref=sr_1_2?keywords=load+bearing+vest&qid=1637893736&sr=8-2), [Load Bearing Vest](https://www.amazon.com/SDS-Official-Military-Fighting-Tactical/dp/B004IURGGY/ref=sr_1_1?keywords=MOLLE+2+Army+Vest&qid=1638131513&sr=8-1), [Heavy Load Bearing Vest](https://www.amazon.com/Loaded-Gear-Tactical-Training-Platforms/dp/B07C2K2FCW/ref=sr_1_3?keywords=load+bearing+vest&qid=1637893736&sr=8-3) ignore the leg pouches
2. Adds flags PALS_SMALL, PALS_MEDIUM and PALS_LARGE representing the amount of space an accessory would take up on a tactical vest
3. Adds 9 new storage containers and gave them appropriate PALS flag: [radio_pouch](https://www.patriotoutfitters.com/5-11-tactical-molle-radio-pouch?PMSRCH=MOLLE%20GEAR), [gadget_pouch](https://www.patriotoutfitters.com/condor-gadget-pouch-tan498-?PMSRCH=MOLLE%20GEAR), [h2o_pouch](https://www.patriotoutfitters.com/condor-h2o-pouch?PMSRCH=MOLLE%20GEAR), [tool_pouch](https://www.patriotoutfitters.com/condor-t-and-t-pouch?PMSRCH=MOLLE%20GEAR), [flashlight_pouch](https://www.patriotoutfitters.com/condor-flashlight-pouch?PMSRCH=MOLLE%20GEAR), [Vertical Pistol Holster tactical_holster](https://www.patriotoutfitters.com/condor-vertical-molle-holster?PMSRCH=MOLLE%20GEAR), [gasmask_pouch](https://condoroutdoor.com/ma11.html), [tactical_grenade_pouch](https://www.tacticalassaultgearstore.com/molle-frag-grenade-pouch.html), [triple_stacker_pouch](https://www.patriotoutfitters.com/condor-triple-stacker-open-top-m4-mag-pouch?PMSRCH=MOLLE%20GEAR), [tactical_shotshell_pouch](https://condoroutdoor.com/ma12-shotgun-ammo-pouch.html)
4. Adds new PALS tags to current items: ifak_pouch, sheath, grenade_pouch, legpouch_large, dump_pouch, holster
5. Distributes the new items on soldiers with separate weightings and item groups. They are also added to mil surplus stores and the generic gun store loot list.
6. Adds new use_actions molle_attach and molle_detach with variable "size" which is an abstract measure of how much can fit on the vest that has the action. This abstraction is used because longest side and volume aren't good measures of the surface area an item takes on a tactical vest since vertical clearance is needed to access pockets and such. 

C++ Code Changes:
item_contents can now track and serialize / de-serialize additional items that are "attached to it" as an array called additional_pockets. This is used for re-adding the dynamic pockets on game load. 
molle_attach can take an item with a PALS flag and incorporate it into a load bearing vest assuming there is still available space for it. These items contribute to the weight and volume of the vest and also give it additional pockets.
molle_detach can remove an incorporated item from a vest taking all the items that were in the pockets it was contributing to the vest into the item and returning it to the player.
Encumbrance is dynamic based on the simple rule of 1 encumbrance per 250ml of storage. 

#### Describe alternatives you've considered
I originally was trying to do this as holsters that take items with specific flags so you don't have a vest with two magazine pockets, you instead have a vest that contains two magazine pouches and those have items in them. This doesn't work well and is visually confusing.

#### Testing
Example Vest:
image

Each item is at the bottom of the page

#### Additional context
Here's some images of UI and stuff:
image

image

image

image

Here's my data spreadsheet I used for this

The Vests | id | Storage | URL
-- | -- | -- | --
Light LBV | light_load_bearing_vest | 3 Slots | https://www.amazon.com/OAREA-Tactical-Assault-Multicam-Multi-Pockets/dp/B07LF8F1P1/ref=sr_1_2?keywords=load+bearing+vest&qid=1637893736&sr=8-2
Medium LBV | load_bearing_vest | 6 Slots | https://www.amazon.com/SDS-Official-Military-Fighting-Tactical/dp/B004IURGGY/ref=sr_1_1?keywords=MOLLE+2+Army+Vest&qid=1638131513&sr=8-1
Heavy LBV | heavy_load_bearing_vest | 10 Slots (covers the stomach too) | https://www.amazon.com/Loaded-Gear-Tactical-Training-Platforms/dp/B07C2K2FCW/ref=sr_1_3?keywords=load+bearing+vest&qid=1637893736&sr=8-3

PALS Compliant new Items | id | Size | URL To add
-- | -- | -- | --
Radio Pouch | radio_pouch | Small (1 slot) | https://www.patriotoutfitters.com/5-11-tactical-molle-radio-pouch?PMSRCH=MOLLE%20GEAR
Gadget Pouch | gadget_pouch | Small (1 slot) | https://www.patriotoutfitters.com/condor-gadget-pouch-tan498-?PMSRCH=MOLLE%20GEAR
H2O Pouch | h2o_pouch | Medium (2 slot) | https://www.patriotoutfitters.com/condor-h2o-pouch?PMSRCH=MOLLE%20GEAR
Tool Pouch | tool_pouch | Large (3 slot) | https://www.patriotoutfitters.com/condor-t-and-t-pouch?PMSRCH=MOLLE%20GEAR
Flashlight Pouch | flashlight_pouch | Small (1 slot) | https://www.patriotoutfitters.com/condor-flashlight-pouch?PMSRCH=MOLLE%20GEAR
Vertical Pistol Holster | tactical_holster | Small (1 slot) | https://www.patriotoutfitters.com/condor-vertical-molle-holster?PMSRCH=MOLLE%20GEAR
Gas Mask Pouch | gas_mask_pouch | Medium (2 slot) | https://condoroutdoor.com/ma11.html
MOLLE Frag Grenade Pouch | tactical_grenade_pouch | Small (1 slot) | https://www.tacticalassaultgearstore.com/molle-frag-grenade-pouch.html
Triple Stacker |  triple_stacker_pouch | Large (3 slot) | https://www.patriotoutfitters.com/condor-triple-stacker-open-top-m4-mag-pouch?PMSRCH=MOLLE%20GEAR
Tactical Shotshell Pouch|  tactical_shotshell_pouch | Medium (2 slot) | https://condoroutdoor.com/ma12-shotgun-ammo-pouch.html

Old Items | id | Size | URL To add
-- | -- | -- | --
IFAK Pouch | ifak_pouch | Medium (2 slot) |  
Sheeth | sheath | Small (1 slot) |  
Grenade Pouch | grenade_pouch | Medium (2 slot) |  
Leg ammo pouch | legpouch_large | Medium (2 slot) |  
Dump Pouch | dump_pouch | Medium (2 slot) | https://www.patriotoutfitters.com/high-speed-gear-mag-net-dump-pouch?PMSRCH=MOLLE%20GEAR
Holster | holster | Medium (2 slot) |  

image
image
image
image
image
image
image
image
image
image
image
image
image

### CODE REVIEW

**Comment by Shodan14** [comment]
Nice, one step closer to Jagged Alliance.

**Comment by bombasticSlacks** [comment]
I'm cleaning up all the clang errors tonight don't worry about reviews

File: data/json/items/armor/ammo_pouch.json [hunk 1]
```diff
@@ -480,6 +480,374 @@
...
+    "material_thickness": 1.5,
+    "use_action": [ { "type": "attach_molle", "size": 3 }, { "type": "detach_molle" } ],
+    "flags": [ "WATER_FRIENDLY", "BELTED" ],
+    "armor": [ { "coverage": 55, "covers": [ "torso" ], "specifically_covers": [ "torso_upper" ] } ]
+  },
+  {
+    "id": "load_bearing_vest",
+    "type": "ARMOR",
+    "name": { "str": "load bearing vest" },
+    "description": "This vest covered in MOLLE webbing is designed to attach pouches, holsters and other PALS compatible accessories.  Prefered by some soldiers over directly attaching accessories to a ballistic vest as they can then be easily seperated.",
```
Comment (Maleclypse) [inline data/json/items/armor/ammo_pouch.json:505]:
```suggestion
    "description": "This vest covered in MOLLE webbing is designed to attach pouches, holsters and other PALS compatible accessories.  Preferred by some soldiers over directly attaching accessories to a ballistic vest as they can then be easily separated.",
```

File: data/json/items/armor/ammo_pouch.json [hunk 2]
```diff
@@ -480,6 +480,374 @@
     "proportional": { "weight": 1.25, "volume": 1.13, "price": 1.25 },
     "flags": [ "WATER_FRIENDLY", "BELTED", "OVERSIZE" ]
   },
+  {
+    "id": "light_load_bearing_vest",
+    "type": "ARMOR",
+    "name": { "str": "light load bearing vest" },
+    "description": "This lightweight harness covered in MOLLE webbing is designed to attach a few pouches, holsters and other PALS compatible accessories.  Prefered by some soldiers over directly attaching accessories to a ballistic vest as they can then be easily seperated.",
```
Comment (Maleclypse) [inline data/json/items/armor/ammo_pouch.json:487]:
```suggestion
    "description": "This lightweight harness covered in MOLLE webbing is designed to attach a few pouches, holsters and other PALS compatible accessories.  Preferred by some soldiers over directly attaching accessories to a ballistic vest as they can then be easily separated.",
```

File: data/json/items/armor/ammo_pouch.json [hunk 3]
```diff
@@ -480,6 +480,374 @@
...
+    "material_thickness": 1.5,
+    "use_action": [ { "type": "attach_molle", "size": 6 }, { "type": "detach_molle" } ],
+    "flags": [ "WATER_FRIENDLY", "BELTED" ],
+    "armor": [ { "coverage": 85, "covers": [ "torso" ], "specifically_covers": [ "torso_upper" ] } ]
+  },
+  {
+    "id": "heavy_load_bearing_vest",
+    "type": "ARMOR",
+    "name": { "str": "heavy load bearing vest" },
+    "description": "This massive vest covered in MOLLE webbing is designed to attach a large number of pouches, holsters and other PALS compatible accessories.  Prefered by some soldiers over directly attaching accessories to a ballistic vest as they can then be easily seperated.",
```
Comment (Maleclypse) [inline data/json/items/armor/ammo_pouch.json:523]:
```suggestion
    "description": "This massive vest covered in MOLLE webbing is designed to attach a large number of pouches, holsters and other PALS compatible accessories.  Preferred by some soldiers over directly attaching accessories to a ballistic vest as they can then be easily separated.",
```

**Review by Maleclypse — COMMENTED** [review]
Very cool.   Is "tripple" with a double p for a brand name?

**Comment by bombasticSlacks** [comment]
Nope I'm just awful at spelling! Thanks for the review.

**Comment by PatrikLundell** [comment]
Spelling: The post has a link to "tripple_stacker_pouch", as well as the same spelling in the PALS list. You may want to update the main post as well (the code is the most important part, though).

**Comment by ISuckM8** [comment]
Hey lad, what are you gonna do with the tac vest? It pretty much functions as a lite version of your load bearing vests.

**Comment by bombasticSlacks** [comment]
The static vests won't change at all. For people that don't want to interact with this system and just want one size fits all options the utility vest, tac vest and others should exist just as they do IRL they also could have better balance (Encumbrance) than a fully custom tac rig. 

As far as I know the reason this isn't on cops is they would more accurately have the one size fits all tac vests. SWAT and Riot cops might use LBVs but their drops in general need an overhaul and I don't want to do it right now. If the New England police should be carrying molle LBVs then someone else could add specifically vests matching that style (or show they are appropriately similar to one of the styles I've added) but again it's research I dont really have the drive to do right now. 

Things I do know and are easy to verify: US soldiers are allowed aftermarket LBVs and mil surplus stores sell piles of cordura pockets and holster with attaching straps in 400 shades of camo

**Comment by Terrorforge** [comment]
Splendid stuff! One small thing: the "grenade pouch" item holds four 40mm grenades, so it's probably more of a size 2. Looks something like this: https://www.hqg.de/en/granatentasche-40mm-4er-lt016-4-2.html

Also, if it's not too much trouble, I would like to request some kind of shotshell holder. Here are a few options I found:

Small strip with a handful of shells: https://www.odsgear.com/product/7-shot-shell-hook-loop-holder-12ga/
Sensible 12-shot pouch: https://condoroutdoor.com/ma12-shotgun-ammo-pouch.html
Low-footprint but awkward-looking tacticool 16-shell folding thing: https://www.amazon.com/Tactical-Shotgun-Shotshell-Buttstock-Carrier/dp/B08R3JKQ89
Redonkulous laptop-sized 25-shell bag: https://www.combathunting.com/Tactical_Vests_Chest_Rigs-Molle_Tactical_Shotgun_Shell_Pouch_Military_Green.html
Slightly less redonkulous but still huge 21-shell placard: https://hrttacticalgear.com/product/hrt-shotgun-placard/
Mag pouch with a few shell loops (may or may not be an airsoft thing): https://m.aliexpress.com/item/32847464205.html?trace=wwwdetail2mobilesitedetail

Really I'm only asking for the basic pouch, but I leave you with the rest as reference in case you wanna go nuts.

**Comment by bombasticSlacks** [comment]
For some reason I thought you would strap the 40mm pouch vertical but that seems like an accident waiting to happen. Since Shotshells would be the same way I've added the "sensible 12 shot pouch". I've also added the additional feature it has of instead holding 20 308 rounds for everyones police sniper rifle shooting needs.

image

**Comment by ISuckM8** [comment]
This is looking great so far!

**Pull request merged**